File: code/scenes/src/main/java/org/fbme/scenes/controllers/scene/CompletionProviderByViewpoint.java
package org.fbme.scenes.controllers.scene;

import org.fbme.scenes.controllers.SceneViewpoint;
import org.fbme.scenes.viewmodel.PositionalCompletionItem;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Supplier;

public class CompletionProviderByViewpoint implements SceneCompletionProvider {

    private final Supplier<List<PositionalCompletionItem>> myCompletionProvider;
    private final SceneViewpoint myViewpoint;

    public CompletionProviderByViewpoint(SceneViewpoint viewpoint, Supplier<List<PositionalCompletionItem>> completionProvider) {
        myViewpoint = viewpoint;
        myCompletionProvider = completionProvider;
    }

    @Override
    public List<SceneCompletionAction> getActions() {
        List<SceneCompletionAction> res = new ArrayList<>();
        for (PositionalCompletionItem completionItem : myCompletionProvider.get()) {
            res.add(new SceneCompletionAction() {
                public String getMatchingText(String pattern) {
                    return completionItem.getMatchingText(pattern);
                }

                public String getDescriptionText() {
                    return completionItem.getDescriptionText();
                }

                public void invoke(String pattern, int x, int y) {
                    int modelX = myViewpoint.translateFromEditorX(x);
                    int modelY = myViewpoint.translateFromEditorY(y);
                    completionItem.invoke(pattern, modelX, modelY);
                }
            });
        }
        return res;
    }
}

File: code/scenes/src/main/kotlin/org/fbme/scenes/controllers/scene/CompletionProviderByViewpoint.kt
package org.fbme.scenes.controllers.scene

import org.fbme.scenes.controllers.SceneViewpoint
import org.fbme.scenes.viewmodel.PositionalCompletionItem
import java.util.function.Supplier

class CompletionProviderByViewpoint(
    private val viewpoint: SceneViewpoint,
    private val completionProvider: Supplier<List<PositionalCompletionItem>>
) : SceneCompletionProvider {
    override val actions: List<SceneCompletionAction>
        get() {
            val res: MutableList<SceneCompletionAction> = ArrayList()
            for (completionItem in completionProvider.get()) {
                res.add(object : SceneCompletionAction {
                    override fun getMatchingText(pattern: String?): String? {
                        return completionItem.getMatchingText(pattern)
                    }

                    override val descriptionText: String?
                        get() = completionItem.descriptionText

                    override fun invoke(pattern: String?, x: Int, y: Int) {
                        val modelX = viewpoint.translateFromEditorX(x)
                        val modelY = viewpoint.translateFromEditorY(y)
                        completionItem.invoke(pattern, modelX, modelY)
                    }
                })
            }
            return res
        }
}
